Sources/Datadog/Core/System/AppStateListener.swift

```diff
@@ -0,0 +1,140 @@
+        )
+        // move final state to upperBound
+        taken.finalDate = range.upperBound
+        // filter changes outside of the range
+        taken.changes = taken.changes.filter { range.contains($0.date) }
+        return taken
+    }
+
+    var foregroundDuration: TimeInterval {
+        var duration: TimeInterval = 0.0
+        var lastActiveStartDate: Date?
+        let allEvents = [initialState] + changes.sorted() + [finalState]
+        for event in allEvents {
+            if let startDate = lastActiveStartDate {
+                duration += event.date.timeIntervalSince(startDate)
+            }
+            if event.isActive {
+                lastActiveStartDate = event.date
+            } else {
+                lastActiveStartDate = nil
+            }
+        }
+        return duration
+    }
+
+    var didRunInBackground: Bool {
+        return !initialState.isActive || !finalState.isActive
+    }
+
+    private func isActive(at date: Date) -> Bool {
+        if date <= initialState.date {
+            // we assume there was no change before initial state
+            return initialState.isActive
+        } else if finalState.date <= date {
+            // and no change after final state
+            return finalState.isActive
+        }
+        var active = initialState.isActive
+        for change in changes {
+            if date < change.date {
+                break
+            }
+            active = change.isActive
+        }
+        return active
+    }
+}
+
+internal protocol AppStateListening {
+    var history: AppStateHistory { get }
+}
+
+internal class AppStateListener: AppStateListening {
+    typealias Snapshot = AppStateHistory.Snapshot
+
+    private static var isActive: Bool {
+        return UIApplication.managedShared?.applicationState == .active
+    }
+
+    private static func currentState(with date: Date) -> Snapshot {
+        return Snapshot(isActive: AppStateListener.isActive, date: date)
+    }
+
+    private let dateProvider: DateProvider
+    private var _appStateHistory: AppStateHistory
+    var history: AppStateHistory {
+        objc_sync_enter(self)
+        defer { objc_sync_exit(self) }
+        _appStateHistory.finalDate = dateProvider.currentDate()
+        return _appStateHistory
+    }
+
+    init(dateProvider: DateProvider = SystemDateProvider()) {
+        self.dateProvider = dateProvider
+        let currentState = AppStateListener.currentState(with: dateProvider.currentDate())
+        _appStateHistory = AppStateHistory(
+            initialState: currentState,
+            changes: [],
+            finalDate: currentState.date
+        )
+        let nc = NotificationCenter.default
+        nc.addObserver(self, selector: #selector(appWillResignActive), name: UIApplication.willResignActiveNotification, object: nil)
+        nc.addObserver(self, selector: #selector(appDidBecomeActive), name: UIApplication.didBecomeActiveNotification, object: nil)
```

Comment:
`didEnterBackground/willEnterForeground` pair wasn't covering all the states, leaving `.inactive` state out.
now i use `willResignActive/didBecomeActive` instead, this gives more accurate `foreground/active duration` information as it properly measures `active` state.